flutter_local_notifications/android/src/main/java/com/dexterous/flutterlocalnotifications/FlutterLocalNotificationsPlugin.java

```diff
@@ -392,24 +396,63 @@ static Gson buildGson() {
   private static ArrayList<NotificationDetails> loadScheduledNotifications(Context context) {
     ArrayList<NotificationDetails> scheduledNotifications = new ArrayList<>();
     SharedPreferences sharedPreferences =
-        context.getSharedPreferences(SCHEDULED_NOTIFICATIONS, Context.MODE_PRIVATE);
-    String json = sharedPreferences.getString(SCHEDULED_NOTIFICATIONS, null);
-    if (json != null) {
+        context.getSharedPreferences(SCHEDULED_NOTIFICATIONS_FILE, Context.MODE_PRIVATE);
+    Set<String> jsonNotifications = getScheduledNotificationsJsonSet(sharedPreferences);
+    if (jsonNotifications != null) {
       Gson gson = buildGson();
-      Type type = new TypeToken<ArrayList<NotificationDetails>>() {}.getType();
-      scheduledNotifications = gson.fromJson(json, type);
+      Type type = new TypeToken<NotificationDetails>() {}.getType();
+      for (String json : jsonNotifications) {
+        scheduledNotifications.add(gson.fromJson(json, type));
+      }
+    }
+    return scheduledNotifications;
+  }
+
+  /**
+   * Get scheduled notifications from shared preferences, converting from old format if present.
+   *
+   * Returns null if neither are present
+   * The returned Set may not be mutated!
+   */
+  private static Set<String> getScheduledNotificationsJsonSet(SharedPreferences sharedPreferences) {
+    Set<String> jsonNotifications = sharedPreferences.getStringSet(SCHEDULED_NOTIFICATIONS_SET, null);
+    if (jsonNotifications != null) {
```

Comment:
This check implies that we cannot switch back to the old implementation and again to this one, the migration will always happen only once